modules/core/src/test/java/org/apache/ignite/internal/processors/cache/consistentcut/AbstractConsistentCutTest.java

```diff
@@ -0,0 +1,296 @@
+import org.apache.ignite.internal.processors.cache.version.GridCacheVersion;
+import org.apache.ignite.internal.util.tostring.GridToStringInclude;
+import org.apache.ignite.internal.util.typedef.G;
+import org.apache.ignite.internal.util.typedef.T2;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.lang.IgniteBiTuple;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.pagemem.wal.record.WALRecord.RecordType.CONSISTENT_CUT_FINISH_RECORD;
+import static org.apache.ignite.internal.pagemem.wal.record.WALRecord.RecordType.CONSISTENT_CUT_START_RECORD;
+import static org.apache.ignite.internal.pagemem.wal.record.WALRecord.RecordType.DATA_RECORD_V2;
+import static org.apache.ignite.internal.processors.cache.persistence.snapshot.AbstractSnapshotSelfTest.snp;
+
+/** Base class for testing Consistency Cut algorithm. */
+public abstract class AbstractConsistentCutTest extends GridCommonAbstractTest {
+    /** */
+    protected static final String CACHE = "CACHE";
+
+    /** */
+    protected static final String SNP = "base";
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String instanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(instanceName);
+
+        cfg.setDataStorageConfiguration(new DataStorageConfiguration()
+            .setWalCompactionEnabled(true)
+            .setDefaultDataRegionConfiguration(new DataRegionConfiguration()
+                .setName("consistent-cut-persist")
+                .setPersistenceEnabled(true)));
+
+        cfg.setCacheConfiguration(cacheConfiguration(CACHE));
+
+        cfg.setConsistentId(instanceName);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        cleanPersistenceDir();
+
+        startGrids(nodes());
+
+        grid(0).cluster().state(ClusterState.ACTIVE);
+
+        startClientGrid(nodes());
+
+        snp(grid(0)).createSnapshot(SNP).get();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
+
+    /** */
+    protected CacheConfiguration<Integer, Integer> cacheConfiguration(String cacheName) {
+        return new CacheConfiguration<Integer, Integer>()
+            .setName(cacheName)
+            .setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL)
+            .setBackups(backups());
+    }
+
+    /**
+     * @return Number of server nodes.
+     */
+    protected abstract int nodes();
+
+    /**
+     * @return Number of backups for cache.
+     */
+    protected abstract int backups();
+
+    /** */
+    protected void awaitSnapshotResourcesCleaned() {
+        try {
+            GridTestUtils.waitForCondition(() -> {
```

Comment:
It seems we need to `assertTrue(waitForCondition())`

waitForCondition don't throw if condition doesn't meet before timeout